nautobot_firewall_models/models/core_models.py

```diff
@@ -656,3 +656,227 @@ def to_json(self):
+        """Return detail view URL."""
+        return reverse("plugins:nautobot_firewall_models:natpolicyrule", args=[self.pk])
+
+    def rule_details(self):
+        """Convenience method to convert to more consumable dictionary."""
+        row = {}
+        row["rule"] = self
+        row["source_users"] = self.source_users.all()
+        row["source_user_groups"] = self.source_user_groups.all()
+        row["source_zone"] = self.source_zone
+        row["destination_zone"] = self.destination_zone
+
+        row["original_source_address_groups"] = self.original_source_address_groups.all()
+        row["original_source_addresses"] = self.original_source_addresses.all()
+        row["original_source_services"] = self.original_source_services.all()
+        row["original_source_service_groups"] = self.original_source_service_groups.all()
+
+        row["translated_source_address_groups"] = self.translated_source_address_groups.all()
+        row["translated_source_addresses"] = self.translated_source_addresses.all()
+        row["translated_source_services"] = self.translated_source_services.all()
+        row["translated_source_service_groups"] = self.translated_source_service_groups.all()
+
+        row["original_destination_address_groups"] = self.original_destination_address_groups.all()
+        row["original_destination_addresses"] = self.original_destination_addresses.all()
+        row["original_destination_services"] = self.original_destination_services.all()
+        row["original_destination_service_groups"] = self.original_destination_service_groups.all()
+
+        row["translated_destination_address_groups"] = self.translated_destination_address_groups.all()
+        row["translated_destination_addresses"] = self.translated_destination_addresses.all()
+        row["translated_destination_services"] = self.translated_destination_services.all()
+        row["translated_destination_service_groups"] = self.translated_destination_service_groups.all()
+
+        row["mode"] = self.mode
+        row["log"] = self.log
+        row["status"] = self.status
+        row["request_id"] = self.request_id
+        return row
+
+    def to_json(self):
+        """Convenience method to convert to json."""
+        return model_to_json(self)
+
+    def __str__(self):
+        """Stringify instance."""
+        if self.request_id and self.name:
+            return f"{self.name} - {self.request_id}"
+        if self.name:
+            return self.name
+        return str(self.id)
+
+
+@extras_features(
+    "custom_fields",
+    "custom_links",
+    "custom_validators",
+    "export_templates",
+    "graphql",
+    "relationships",
+    "statuses",
+    "webhooks",
+)
+class NATPolicy(PrimaryModel):
+    """
+    The overarching model that is the full NAT policy with all underlying rules and child objects.
+
+    Each NATPolicy can be assigned to both devices and to dynamic groups which in turn can assign the policy to a related device.
```

Comment:
Is there a reason we're only applying these to the whole Device and not to individual interfaces? I would think end users would find it more appropriate for a NAT policy to be applied to an interface more than a whole Device.

Reply:
Couple of thoughts here:
- I am following the established pattern for security rules, so consistency is a plus in my opinion
- Check Point for example has NAT rules applied to entire devices matching on routing lookups IIRC, so not all vendors implement NAT rules directly to interfaces

Reply:
I feel like we should have both NAT and Security policies be able to be applied to the whole device and to individual interfaces. Think of the case of a router where the ACLs are applied to a particular interface and in a particular direction. For ASA all ACLs are applied typically to an interface. The same goes for NATs. You are correct that some act globally but in general I think the rule is to apply to an interface.